### PR TITLE
Implement rich-presence for players in matches

### PR DESCRIPTION
**Pull request by Liam-Directive**
This PR must be paired with the following PRs:
 - Perseus: https://github.com/directivegames/perseus/pull/90
 - DriftBase UE4: https://github.com/directivegames/DriftUE4Base/pull/38

This PR introduces the concept of 'Rich Presence' for the player, so that it's possible to track live-updated information about your friends, such as the current map they're playing, or whether they're offline.

This information can be displayed in-game, for example in the lobby friends list.

In a nutshell, this information is available in two ways:
1) By request, via a dedicated endpoint
2) By listening to 'richpresence' queue using the messaging system

The information itself is kept fresh as things change rather than by polling. So for example when a player leaves a match, this will be detected, and his rich presence information will be updated to reflect that.

## Updates 09/10:

I've updated the `richpresence` service to be a class with dependency injection (e.g., db, redis). I've also purified the flask implementation so that it purely relies on the service, and doesn't do any of it's own business logic. 

I also updated from using a document-store for all rich presence information, to instead use a hashset. I'm not positive whether my implementation is correct. Do I still need a lock?

### CODE REVIEW

File: driftbase/models/db.py [hunk 1]
```diff
@@ -114,7 +112,35 @@ def check_password(self, password):
...
+        try:
+            match_player : MatchPlayer = g.db.query(MatchPlayer) \
+                .filter(MatchPlayer.player_id == player_id) \
+                .one()
+            
+            match : Match = g.db.query(Match) \
+                .filter(Match.match_id == match_player.match_id) \
+                .one()
+
+            self.is_in_game = match_player.status == "active" # Always true?
```
Comment (judgeaxl) [inline driftbase/models/db.py:138]: Status is update a bit more frequently when using the legacy match maker I believe, but could be "banned" in the future.

File: driftbase/models/db.py [hunk 2]
```diff
@@ -114,7 +112,35 @@ def check_password(self, password):
...
+            self.game_mode = match.game_mode
+            self.map_name = match.map_name
+        except Exception: # Expected NoResultFound if a player isn't in a match.
```
Comment (judgeaxl) [inline driftbase/models/db.py:141]: Please use a more specific exception, so that actual errors can be dealt with differently from not found

File: driftbase/models/db.py (hunk 1 above)
Comment (judgeaxl) [reply driftbase/models/db.py:138]: The actual valid states that are used may be unknown, since it can be set by the battle servers too, I believe

File: driftbase/models/db.py [hunk 3]
```diff
@@ -356,8 +389,10 @@ class Match(ModelBase):
     status_date = Column(DateTime, nullable=True)
     unique_key = Column(String(50), nullable=True)
 
-
 class MatchPlayer(ModelBase):
+    """
+    The match that a player is currently in
```
Comment (judgeaxl) [inline driftbase/models/db.py:394]: This comment is incorrect, one entry is created for every player in every match, but only one with a given match/player combo should exist.

File: driftbase/models/db.py [hunk 4]
```diff
@@ -143,11 +169,18 @@ def __marshallable__(self):
...
     @hybrid_property
     def is_online(self):
+        """
+        Deprecated. Use rich_presence.is_online instead.
+        """
         if self.user and self.user.client:
             return self.user.client.is_online
         return False
 
+    @hybrid_property
```
Comment (judgeaxl) [inline driftbase/models/db.py:179]: I'm not convinced that putting arbitrary queries for related data into the core player is the way to go, especially since it uses `g.db` directly, which makes it harder to mock and test this logic. Admittedly the layering in drift-base is very poor, so this may not be a real-world issue until we rewrite the whole thing, but I still feel that it would make more sense to query for this explicitly where it is needed.

If there was a relationship between core players and match players perhaps, but I don't think such a relationship should be on the model, since they are unrelated concepts, and candidates for being refactored into separate applications.

File: driftbase/models/db.py [hunk 5]
```diff
@@ -114,7 +112,35 @@ def check_password(self, password):
 
 tbl_user_identity = UserIdentity.__table__
 
-
+class PlayerRichPresence:
+    """
+    Rich presence information for a particular user.
```
Comment (judgeaxl) [inline driftbase/models/db.py:117]: Please don't mix user and player terminology, they are not the same. This would be a player.

File: tests/players/test_players.py [hunk 6]
```diff
@@ -50,6 +50,44 @@ def test_players_online(self):
...
+        matchplayers_url = resp["matchplayers_url"]
+
+        resp = self.post(teams_url, data={}, expected_status_code=http_client.CREATED).json()
+        team_id = resp["team_id"]
+        resp = self.get(teams_url).json()
+
+        data = {"player_id": player_id, "team_id": team_id}
+        resp = self.post(matchplayers_url, data=data, expected_status_code=http_client.CREATED).json()
+        
+        presence = self._get_player_rich_presence()
```
Comment (judgeaxl) [inline tests/players/test_players.py:85]: Please add tests for the cases where this data does not exist, or map_name and game_mode are empty or None.

File: tests/players/test_players.py [hunk 7]
```diff
@@ -50,6 +50,44 @@ def test_players_online(self):
...
+        self.assertEqual(presence["is_online"], self.get(self.endpoints["my_player"]).json()["is_online"])
+        self.assertTrue(presence["is_in_game"])
```
Comment (judgeaxl) [inline tests/players/test_players.py:87]: Since this is depending on the status field, you'll want to test that it reflects the status correctly.

**Review by judgeaxl — CHANGES_REQUESTED** [review]
A good addition, but I'm not convinced that a hybrid_property is the best fit

File: driftbase/models/db.py (hunk 4 above)
Comment (Liam-Directive) [reply driftbase/models/db.py:179]: Thanks for the feedback! 

Just to clarify my reasoning: In Perseus we access `is_online` field by querying for friends, and then filtering only for the 'is_online' column. We poll these values every few seconds. This existing setup is what inspired me to stick rich_presence into the player, so we can keep the logic in Perseus largely the same. 

**Proposed alternative:** I can introduce a completely separate flask endpoint called something like `get_rich_presence`, and return the `g.db` calculated data directly. This also allows me to toss some boilerplate related to Marshmallow (at least I think so). Thoughts?

File: driftbase/models/db.py (hunk 5 above)
Comment (Liam-Directive) [reply driftbase/models/db.py:117]: Thanks. I will remove these doc-strings for now to avoid incorrect documentation.

File: driftbase/models/db.py (hunk 1 above)
Comment (Liam-Directive) [reply driftbase/models/db.py]: Yeah we may need to keep an eye on this "status" field, and whether it's actually a reliable source of information for the matches status. I think for now at least, 'active' is reliably used when the match is ongoing.

File: driftbase/models/db.py (hunk 4 above)
Comment (Liam-Directive) [reply driftbase/models/db.py]: Implemented something along the lines of what I described. If you have a different design idea, just let me know :)

Currently, what's missing from my implementation is an easy way to grab the status for a lot of related players. We used to get this for "free" from the players endpoint, since we could filter players by friend status, and then filter also the db rows returned by key. 

Now that I moved rich presence into it's own endpoint, we lack this flexibility. I can add something like `richpresence.list` (see `players.list`), but it would require me to re-implement the filtering options defined there. Not too bad to do, but I wanted to hear your opinions on the direction first.

File: driftbase/api/players/__init__.py [hunk 8]
```diff
@@ -235,9 +235,17 @@ def endpoint_info(current_user):
...
+        "template_player_gamestate": template_player_gamestate_url
     }
 
+    url = url_for(
+        "richpresence.entry",
+        player_id=1337,
+        _external=True,
+    )
+    url = url.replace('1337', '{player_id}')
+    ret["rich_presence"] = url
```
Comment (Liam-Directive) [inline driftbase/api/players/__init__.py:247]: Self review: This probably should be called `template_richpresence` to match the other templated URLs, right?

File: driftbase/api/players/richpresence.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+class RichPresenceAPI(MethodView):
+    @bp.response(http_client.OK, RichPresenceResponseSchema)
+    def get(self, player_id : int):
+        """
+        Single Player
+
+        Retrieve rich-presence information for a specific player
+        """
+                
+        return PlayerRichPresence(player_id)
```
Comment (judgeaxl) [inline driftbase/api/players/richpresence.py:69]: You need to handle any exceptions and turn them into the proper status codes, 200, 400, 404, 500, etc.

File: driftbase/api/players/richpresence.py [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+from drift.core.extensions.urlregistry import Endpoints
+from flask import g, url_for
+import logging
+
+
+bp = Blueprint("richpresence", __name__, url_prefix="/players/<int:player_id>")
+endpoints = Endpoints()
+log = logging.getLogger(__name__)
+
+class RichPresenceRequestSchema(Schema):
```
Comment (judgeaxl) [inline driftbase/api/players/richpresence.py:16]: This looks unused

File: driftbase/api/players/__init__.py (hunk 8 above)
Comment (judgeaxl) [reply driftbase/api/players/__init__.py]: Yes

File: driftbase/models/db.py (hunk 4 above)
Comment (judgeaxl) [reply driftbase/models/db.py]: You could have root resources /[rich_]presence[?player_ids=[...]] and /[rich_]presence/player-id since this could be thought of as its own resource. 

It does seem odd to have the client ask for this every few seconds though. The pattern ought to be:

1. Ask for the current state of all relevant players
2. Listen for updates on the messages endpoint

I think the notifications are probably not being sent to all relevant players now, and that's what's preventing this from being as nice as it could be, so looking into how that can be improved, if it's possible or won't work for all cases, would be a good idea.

But yes, having a way to query for more than one player at a time is definitely worth it compared to bombarding the backend with multiple requests. In fact, Unreal is already complaining that we're sending too many requests in parallel

Comment (Liam-Directive) [reply driftbase/models/db.py]: Richpresence has been migrated into it's own flask extension. 

Additionally, rich presence is no longer *polled* at all, but rather kept up to date in redis by listening to various events (e.g., player join match, player leave match). 

It's now possible to query information (via flask endpoint), or listen to the message queue 'richpresence' in order to get updates without polling.

**Comment by judgeaxl** [comment]
There's one thing here that while it mimics how a lot of code in drift-base is already written, goes against how we want new code to be written. So instead of having lose methods that go through `g` or other global objects and functions to find dependencies, we build service classes that take those dependencies as inputs. This vastly simplifies testing and ensures one don't start calling those methods outside a context where the dependencies are available. You can check most other apps, except seasons, for how this pattern is used.

File: driftbase/api/messages.py [hunk 11]
```diff
@@ -210,7 +210,16 @@ def get(self, exchange, exchange_id, queue, message_id):
 
 @endpoints.register
 def endpoint_info(*args):
-    return {
+    # TODO: This way of building templated URLs is very repetitive. I'm positive we can wrap this.
```
Comment (judgeaxl) [inline driftbase/api/messages.py:213]: Who will use this template? The system is actually quite restrictive on who can send things to each queue and exchange, so it might not be all that useful outside the pre-setup my_messages URL

File: driftbase/api/richpresence.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+class RichPresenceAPI(MethodView):
+    @bp.response(http_client.OK, RichPresenceSchema)
+    def get(self, player_id : int):
+        """
+        Single Player
+
+        Retrieve rich-presence information for a specific player
+        """
+        
+        player = g.db.query(CorePlayer).get(player_id)
```
Comment (judgeaxl) [inline driftbase/api/richpresence.py:46]: I'd suggest leaving the player check to the service code, throw an exception and turn it into a 404 here, instead of having partial logic in the API handler. drift-base does not follow the rules very well, but API handlers should really only concern themselves with validating the input follows approved syntax, and formatting the responses, they should have no business logic in them. Again, not a pattern drift-base follows in general, but since this is new code, why not.

File: driftbase/richpresence.py [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+def _get_redis_key(player_id) -> str:
+    return g.redis.make_key(f"rich_presence:{player_id}")
+    
+def get_richpresence(player_id : int) -> PlayerRichPresence:
+    """
+    Fetches the rich presence information for the specified player.
+    """
+
+    key = _get_redis_key(player_id)
+    presence_json = g.redis.get(key)
```
Comment (judgeaxl) [inline driftbase/richpresence.py:61]: When you use the methods on the redis instance, they will scope the key internally using `g.redis.make_key()`. If you pass in a pre-scoped key from `make_key()` you'll double scope it. Either use `g.redis.func()` with un-scoped keys, or use `g.redis.conn.func()` with pre-scoped keys.

File: driftbase/richpresence.py [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+    if presence_json:
+        return RichPresenceSchema(many=False).load(presence_json)
+    else:
+        return PlayerRichPresence()
+
+def set_online_status(player_id: int, is_online: bool):
+    """
+    Sets the players online status, and updates rich-presence
+    """
+    presence = get_richpresence(player_id)
```
Comment (judgeaxl) [inline driftbase/richpresence.py:72]: Read-modify-write is a candidate for races if multiple instances try to update the rich presence at the same time. You need a lock, or at the very least a way to modify individual keys atomically. Instead of using get/set with serialized JSON, consider using a hash per player where keys can be modified individually. Races may still overwrite values out of order, but at least not the entire state.

Comment (judgeaxl) [reply driftbase/richpresence.py:72]: see hset, hget, hgetall

File: driftbase/api/clients.py [hunk 15]
```diff
@@ -342,6 +347,8 @@ def delete(self, client_id):
 
         current_app.extensions["messagebus"].publish_message("client", message_data)
 
+        richpresence.set_online_status(player_id, False)
```
Comment (judgeaxl) [inline driftbase/api/clients.py:350]: You want to ensure this never throws, or wrap it and ignore failures, since this is a side effect and not a requirement for the success of DELETE /clients/n

File: driftbase/api/clients.py [hunk 16]
```diff
@@ -235,6 +237,7 @@ def post(self, args):
 
         current_app.extensions["messagebus"].publish_message("client", message_data)
 
+        richpresence.set_online_status(player_id, True)
```
Comment (judgeaxl) [inline driftbase/api/clients.py:240]: You want to ensure this never throws, or wrap it and ignore failures, since this is a side effect and not a requirement for the success of POST /clients

File: driftbase/api/matches.py [hunk 17]
```diff
@@ -814,6 +815,9 @@ def post(self, args, match_id):
                                        seconds=0,
                                        status="active")
             g.db.add(match_player)
+        
+        richpresence.set_match_status(player_id, match.map_name, match.game_mode)
```
Comment (judgeaxl) [inline driftbase/api/matches.py:819]: Wrap try/catch to ensure rich presence don't break match management

File: driftbase/api/matches.py [hunk 18]
```diff
@@ -944,6 +948,9 @@ def delete(self, match_id, player_id):
         team_id = match_player.team_id
 
         match_player.status = "quit"
+
+        richpresence.clear_match_status(player_id)
```
Comment (judgeaxl) [inline driftbase/api/matches.py:952]: Wrap try/catch to ensure rich presence don't break match management

File: driftbase/api/richpresence.py [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+class RichPresenceListArgs(Schema):
+    class Meta:
+        strict = True
+
+    player_id = fields.List(
+        fields.Integer(), metadata=dict(description="Player ID's to filter for"
+    ))
+
+@bp.route('/<int:player_id>', endpoint='entry')
+class RichPresenceAPI(MethodView):
```
Comment (judgeaxl) [inline driftbase/api/richpresence.py:37]: Could there be a reason for not allowing you to query for arbitrary players that you're not friends with or in a match with?

File: driftbase/api/richpresence.py [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+
+        try:
+            return get_richpresence(player_id)
+        except Exception:
+            abort(http_client.INTERNAL_SERVER_ERROR)
+
+
+
+@endpoints.register
+def endpoint_info(*args):
```
Comment (judgeaxl) [inline driftbase/api/richpresence.py:58]: The rich presence url for a player could be made part of the player info endpoint response

File: driftbase/richpresence.py [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+    presence.is_online = is_online
+    set_richpresence(player_id, presence)
+
+def set_match_status(player_id: int, map_name: str, game_mode: str, is_in_game : bool = True):
+    """
+    Sets the players match status, and updates rich-presence
+    """
+    presence = get_richpresence(player_id)
+    presence.is_in_game = is_in_game
```
Comment (judgeaxl) [inline driftbase/richpresence.py:81]: What's the difference between `is_in_game == True` and `game_mode != "" and map_name != ""`?

File: driftbase/richpresence.py [hunk 22]
```diff
@@ -0,0 +1,107 @@
...
+    Returns an array of player_ids, matching your friends list.
+    """
+
+    left = g.db.query(Friendship.id, Friendship.player1_id, Friendship.player2_id).filter_by(player1_id=player_id, status="active")
+    right = g.db.query(Friendship.id, Friendship.player2_id, Friendship.player1_id).filter_by(player2_id=player_id, status="active")
+    friend_rows = left.union_all(right)
+
+    return [row[2] for row in friend_rows]
+
+def _get_redis_key(player_id) -> str:
```
Comment (judgeaxl) [inline driftbase/richpresence.py:52]: See other comment about using make_key on the RedisCache instance vs on the RedisCache.conn methods

**Review by judgeaxl — CHANGES_REQUESTED** [review]

File: driftbase/api/messages.py (hunk 11 above)
Comment (einarth) [reply driftbase/api/messages.py:213]: I may be missing something, but I'm not getting how this is relevant except for convenience in tests-writing.  Clients will only ever get from the 'players' exchange and only by parsing the messages from that can they get a specific message from a particular queue.

File: driftbase/flask/flaskproxy.py [hunk 23]
```diff
@@ -0,0 +1,22 @@
...
+    """
+
+    # Note: Some of these are actually 'LocalProxy[T]', but for improved type hinting we will hint the underlying type.
+    redis : RedisCache
+    db : Session
+
+    def __getattr__(self, key):
+        return getattr(flask.g, key)
+    def __setattr__(self, name: str, value: Any) -> None:
+        flask.g.__setattr__(name, value)
```
Comment (einarth) [inline driftbase/flask/flaskproxy.py:20]: Any particular reason for not going through setattr(flask.g, name, value) here?

File: driftbase/richpresence.py [hunk 24]
```diff
@@ -0,0 +1,107 @@
...
+class PlayerRichPresence:
+    """
+    Rich presence information for a particular player.
+    @see PlayerRichPresenceSchema
+    """
+    
+    game_mode = ""
+    map_name = ""
+    is_online = False
+    is_in_game = False
```
Comment (einarth) [inline driftbase/richpresence.py:17]: Do these serve any purpose?

File: driftbase/richpresence.py [hunk 25]
```diff
@@ -0,0 +1,107 @@
...
+        return PlayerRichPresence(**data)
+
+def _get_friends(player_id) -> list[int]:
+    """
+    Returns an array of player_ids, matching your friends list.
+    """
+
+    left = g.db.query(Friendship.id, Friendship.player1_id, Friendship.player2_id).filter_by(player1_id=player_id, status="active")
+    right = g.db.query(Friendship.id, Friendship.player2_id, Friendship.player1_id).filter_by(player2_id=player_id, status="active")
+    friend_rows = left.union_all(right)
```
Comment (einarth) [inline driftbase/richpresence.py:48]: refactoring opportunity with FriendshipAPI :)

File: driftbase/utils/test_utils.py [hunk 26]
```diff
@@ -82,6 +89,27 @@ def _create_server(self, machine_id):
...
+                kwargs["path"] = "/test/endpoint"
+            with self.drift_app.test_request_context(*args, **kwargs) as request_ctx:
+                g.conf = LocalProxy(get_config_for_request)
+                g.db = LocalProxy(get_sqlalchemy_session)
+                g.redis = LocalProxy(get_redis_session)
+                try:
+                    yield request_ctx
+                finally:
+                    gevent.idle()
+
```
Comment (einarth) [inline driftbase/utils/test_utils.py:112]: You don't need these; all your tests go through an api, so you don't need to fake going through an api.

File: tests/test_richpresence.py [hunk 27]
```diff
@@ -0,0 +1,135 @@
...
+            self.assertEqual(res['map_name'], "")
+            self.assertEqual(res['game_mode'], "")
+            self.assertFalse(res['is_in_game'])
+
+    def test_richpresence_messagequeue(self):
+        """
+        Tests that updating rich-presence will send a message to your friends.
+        """
+        
+        self.auth()
```
Comment (einarth) [inline tests/test_richpresence.py:112]: superfluous auth/player creation

File: driftbase/flask/flaskproxy.py (hunk 23 above)
Comment (Liam-Directive) [reply driftbase/flask/flaskproxy.py:20]: I would want to confirm how that interacts with the wrapping LocalProxy, but probably it's equivalent. Will check when I'm dealing with the rest of the review.

File: driftbase/richpresence.py (hunk 24 above)
Comment (Liam-Directive) [reply driftbase/richpresence.py:17]: Can you clarify what you mean? Are you suggesting we drop the two bools from the rich presence system?

Comment (einarth) [reply driftbase/richpresence.py:17]: I mean why have those class variables.

Comment (Liam-Directive) [reply driftbase/richpresence.py:17]: Ah yes, you're right. Probably a leftover from using a @dataclass.

File: driftbase/utils/test_utils.py (hunk 26 above)
Comment (Liam-Directive) [reply driftbase/utils/test_utils.py:112]: I need it to call `rp.set_richpresence(friend_id, presence)` in one of the tests. I will revisit this topic after refactoring the service to be better encapsulated from the flask extension.

Comment (Liam-Directive) [reply driftbase/utils/test_utils.py:112]: Note: These context managers already existed. I just moved them "up" one level so they're available in the test utils and usable across all tests, rather than locked to the shoutout implementation.

File: driftbase/api/richpresence.py (hunk 12 above)
Comment (Liam-Directive) [reply driftbase/api/richpresence.py:46]: I understand what you mean. What's the best way of percolating an error from the service to the API handler?  Let's say I want to return `NOT_FOUND` in some cases, instead of always `INTERNAL_SERVER_ERROR`. 

I can either return a tuple of err/value, or throw+handle exceptions? Any preferences?

File: driftbase/api/richpresence.py (hunk 20 above)
Comment (Liam-Directive) [reply driftbase/api/richpresence.py:55]: Done, I guess.

File: driftbase/api/messages.py (hunk 11 above)
Comment (Liam-Directive) [reply driftbase/api/messages.py:213]: I will remove the template. You're correct that it's only really relevant for testing.

Comment (Liam-Directive) [reply driftbase/api/messages.py:213]: Done. I moved the `url_for` stuff into the test harness. Are we OK with using context handlers in this way, or should I be building the URL somehow else?

File: driftbase/richpresence.py (hunk 21 above)
Comment (judgeaxl) [reply driftbase/richpresence.py]: What I meant is, you should not store data that can be calculated on the fly, unless the calculation is expensive, as otherwise you may introduce inconsistencies, .e.g is_in_game could be skipped in storage and be the result of the above check instead

Comment (Liam-Directive) [reply driftbase/richpresence.py]: I think it's fine to compute it when requested, rather than storing it in Redis. I will go with that for now.

File: driftbase/richpresence.py (hunk 25 above)
Comment (Liam-Directive) [reply driftbase/richpresence.py]: Left a FIXME for now. I don't want to muddy the PR with a refactor of the Friendship API.

File: driftbase/utils/test_utils.py (hunk 26 above)
Comment (einarth) [reply driftbase/utils/test_utils.py:112]: We should probably move these further up to drift's systesthelper at some point

**Review by einarth — APPROVED** [review]

File: driftbase/api/richpresence.py (hunk 19 above)
Comment (Liam-Directive) [reply driftbase/api/richpresence.py:36]: Protected it and added tests.
